File: windows/Server/ProxyRequestData.cs
using Newtonsoft.Json;
using System.Collections.Generic;
using System.Text;

namespace Musiche.Server
{
    public class ProxyRequestData
    {
        public string Url { get; set; } = string.Empty;
        public bool SetCookieRename { get; set; } = false;
        private string _method = "GET";
        public string Method
        {
            get { return _method; }
            set
            {
                if (string.IsNullOrWhiteSpace(value)) _method = "GET";
                else _method = value;
            }
        }
        public string Data { get; set; } = string.Empty;
        private byte[] _data = null;
        public byte[] DataBytes
        {
            get
            {
                if (_data == null) _data = Encoding.UTF8.GetBytes(Data);
                return _data;
            }
        }
        public bool HasBody => DataBytes.Length > 0;
        public Dictionary<string, string> Headers { get; set; } = new Dictionary<string, string>();
        public bool AllowAutoRedirect { get; set; } = true;

        public static ProxyRequestData Parse(string data)
        {
            if (data.StartsWith("http"))
            {
                return new ProxyRequestData() { Url = data };
            }
            try
            {
                return JsonConvert.DeserializeObject<ProxyRequestData>(data) ?? new ProxyRequestData();
            }
            catch (System.Exception)
            {
                Logger.Logger.Warning("ProxyRequestData Parse Error");
                return new ProxyRequestData();
            }
        }
    }
}

File: mobile/lib/server/proxy_request_data.dart
import 'dart:convert';

import 'package:musiche/log/logger.dart';

class ProxyRequestData {
  static const String _tag = 'MusicheProxyRequestData';
  String url = "";
  bool setCookieRename = false;
  bool allowAutoRedirect = true;
  Map<String, dynamic>? headers;
  String method = "GET";
  String data = "";
  bool get hasBody => data.isNotEmpty;

  ProxyRequestData();

  factory ProxyRequestData.fromString(String source) {
    ProxyRequestData result = ProxyRequestData();
    if (source.startsWith("http")){
      result.url = source;
      return result;
    }
    Map<String, dynamic>? json;
    try {
      json = jsonDecode(source);
    }catch(e){
      Logger.e(_tag, "parse proxy request data err: $e");
    }
    json?.forEach((key, value) {
      switch(key.toLowerCase()){
        case "url":
          if(value.runtimeType == String) {
            result.url = value;
          }
          break;
        case "setcookierename":
          if(value.runtimeType == bool) {
            result.setCookieRename = value;
          }
          break;
        case "allowautoredirect":
          if(value.runtimeType == bool) {
            result.allowAutoRedirect = value;
          }
          break;
        case "headers":
          if(value is Map<String, dynamic>) {
            result.headers = value.map((key, value) {
              return MapEntry(key, value?.toString() ?? "");
            });
          }
          break;
        case "method":
          if(value.runtimeType == String && value.toString().isNotEmpty) {
            result.method = value;
          }
          break;
        case "data":
          if(value.runtimeType == String) {
            result.data = value;
          }
          break;
      }
    });
    return result;
  }
}